### PR TITLE
feat: build infra

### PR DESCRIPTION
**Pull request by AbhishekA1509**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] Does this PR require documentation updates?
- [ ] I've updated documentation as required by this PR.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/Pages/GlobalConfigurations/BuildInfra/constants.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        LIMIT_DECIMAL_PLACES: 'Limit should be upto 2 decimal places.',
+    },
+    getInvalidActionMessage: (action: HandleProfileInputChangeType['action']) => `Invalid action type: ${action}`,
+    getSubmitSuccessMessage: (profileName: string, isEdited: boolean) =>
+        `${profileName} is ${isEdited ? 'updated' : 'created'}`,
+    PROFILE_NOT_FOUND: {
+        title: 'Profile not found',
+        subTitle: 'The profile you are looking for does not exist.',
+    },
+}
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/constants.tsx:47]: as const for all

File: src/Pages/GlobalConfigurations/BuildInfra/constants.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        locator: BuildInfraLocators.BUILD_TIMEOUT,
+    },
+]
+
+// TODO: Handle the case of create profile to disable save button initially and enable it when all the fields are filled
+export const PROFILE_INPUT_ERROR_FIELDS = {
+    ...Object.fromEntries(
+        Object.keys({ ...BuildInfraConfigTypes, ...BuildInfraMetaConfigTypes }).map((key) => [key, null]),
+    ),
+} as ProfileInputErrorType
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/constants.tsx:113]: remove todo and top level {} and ...

File: src/Pages/GlobalConfigurations/BuildInfra/BuildInfraFormAction.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        <div className="flexbox-col dc__gap-4 dc__mxw-420 w-100 dc__align-start">
+            <label
+                htmlFor={`${actionType}-input`}
+                className={`fs-13 fw-4 lh-20 cn-7 ${isRequired ? 'dc__required-field' : ''}`}
+            >
+                {label}
+            </label>
+            <div className="w-100 flexbox dc__align-items-center">
+                <div className="flex-grow-1">
+                    <input
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/BuildInfraFormAction.tsx:63]: discuss regarding CustomInput

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+}: ValidateRequestLimitType): ValidateRequestLimitResponseType => {
+    // Request <= Limit
+    // Request and Limit should be numbers can be decimals
+    // Request and Limit should be positive numbers
+    // Request and Limit should be less than Number.MAX_SAFE_INTEGER
+    // Request and Limit should can have at most BUILD_INFRA_INPUT_CONSTRAINTS.DECIMAL_PLACES
+    // Both request and limit should be there
+    const requestNumber = Number(request.value)
+    const limitNumber = Number(limit.value)
+    const requestUnit = Number(unitsMap[request.unit].conversionFactor)
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:44]: can it be undefined?

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+            isValid: false,
+        }
+
+        return requestLimitValidationResponse
+    }
+
+    const limitRequestUnitFactor = limitUnit / requestUnit
+    const limitRequestFactor = limitNumber / requestNumber
+
+    const isProductSafe = limitRequestUnitFactor * limitRequestFactor <= Number.MAX_SAFE_INTEGER
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:101]: better name

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+        }
+
+        setProfileInput(currentInput)
+        setProfileInputErrors(currentInputErrors)
+    }
+
+    const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
+        e.preventDefault()
+        setLoadingActionRequest(true)
+        try {
```
Comment (AbhishekA1509) [inline src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:397]: check if validation is required here

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        profile?.configurations?.reduce((acc, configuration) => {
+            acc[configuration.key] = configuration
+            return acc
+        }, {}) ?? {}
+
+    // traversing defaultConfigurationsMap and if key is present in profileConfigurations then use that else use defaultConfigurationsMap
+    const configurations = Object.keys(defaultConfigurationsMap).reduce((acc, key) => {
+        const defaultConfiguration: BuildInfraConfigurationType = defaultConfigurationsMap[key]
+        const profileConfiguration = profileConfigurations[key]
+        if (fromCreateView) {
```
Comment (vivek-devtron) [inline src/Pages/GlobalConfigurations/BuildInfra/services.tsx:66]: We can handle this in else part

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        type: profileInput.type,
+        configurations,
+    }
+    return payload
+}
+
+export const updateBuildInfraProfile = ({ name, profileInput }: UpdateBuildInfraProfileType) => {
+    const payload = getBuildInfraProfilePayload(profileInput)
+    // TODO: Would remove this
+    console.log(payload, 'edit payload')
```
Comment (vivek-devtron) [inline src/Pages/GlobalConfigurations/BuildInfra/services.tsx:156]: Please remove this and other console.log

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    return put(`${ROUTES.INFRA_CONFIG_PROFILE}/${name}`, payload)
+}
+
+export const createBuildInfraProfile = async ({ profileInput }: CreateBuildInfraProfileType) => {
+    // Adding a timeout to show the loader
+    // eslint-disable-next-line no-promise-executor-return
+    await new Promise((resolve) => setTimeout(resolve, 1000))
```
Comment (vivek-devtron) [inline src/Pages/GlobalConfigurations/BuildInfra/services.tsx:163]: Please remove this line

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+            setProfileInput({
+                ...profileResponse.profile,
+            })
+
+            setProfileInputErrors(getInitialProfileInputErrors(fromCreateView))
+        }
+    }, [profileResponse, isLoading])
+
+    // NOTE: Currently sending and receiving values as string, but will parse it to number for payload
+    const handleProfileInputChange = ({ action, data }: HandleProfileInputChangeType) => {
```
Comment (vivek-devtron) [inline src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:171]: Extract out the cases in smaller and common functions.

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx (hunk 9 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/services.tsx]: Done

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx (hunk 8 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/services.tsx]: Done

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx (hunk 6 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:410]: Added check for error field object length

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx (hunk 5 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/utils.tsx]: Renamed to isDifferenceBigNumber

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx (hunk 4 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/utils.tsx]: Should not be possible, but on safer side have added a check.

File: src/Pages/GlobalConfigurations/BuildInfra/BuildInfraFormAction.tsx (hunk 3 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/BuildInfraFormAction.tsx:63]: Have discussed, there was some alignment related issue

File: src/Pages/GlobalConfigurations/BuildInfra/constants.tsx (hunk 2 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/constants.tsx]: Done

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+const getBuildInfraProfilePayload = (
+    profileInput: CreateBuildInfraProfileType['profileInput'],
+): CreateBuildInfraServicePayloadType => {
+    const currentConfigurations = profileInput.configurations
+    const configurationKeys = Object.keys(currentConfigurations) as BuildInfraConfigTypes[]
+    // would only keep the configurations with id or active flag as true
+    const configurations: CreateBuildInfraServiceConfigurationType[] = configurationKeys.reduce((acc, key) => {
+        const configuration = currentConfigurations[key]
+        if (configuration.id || configuration.active) {
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/services.tsx:136]: Working

File: src/Pages/GlobalConfigurations/BuildInfra/types.tsx [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    ACTIVATE_CPU = `activate_cpu`,
+    DE_ACTIVATE_CPU = `de_activate_cpu`,
+    ACTIVATE_MEMORY = `activate_memory`,
+    DE_ACTIVATE_MEMORY = `de_activate_memory`,
+    ACTIVATE_BUILD_TIMEOUT = `activate_build_timeout`,
+    DE_ACTIVATE_BUILD_TIMEOUT = `de_activate_build_timeout`,
+}
+
+// TODO: Better name would be BuildInfraConfigActionMap
+export const BuildInfraLocatorsActionMap = {
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/types.tsx]: Done

File: src/Shared/Components/APIResponseHandler/APIResponseHandler.tsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+    if (isLoading) {
+        return <Progressing {...progressingProps} />
+    }
+
+    if (error) {
+        if (error?.code === 404) {
+            return (
+                <GenericEmptyState
+                    title={notFoundText?.title ?? 'Data not found'}
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Shared/Components/APIResponseHandler/APIResponseHandler.tsx]: Done

File: src/Shared/Components/APIResponseHandler/types.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
+import { ReactNode } from 'react'
+import { ProgressingProps, ReloadType, ServerErrors } from '../../../Common'
+
+export interface EmptyPageTextType {
+    title?: string
+    subTitle?: string
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Shared/Components/APIResponseHandler/types.tsx]: Done

File: src/Shared/validations.tsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            message: 'Maximum 300 characters are allowed',
+        }
+    }
+
+    return {
+        isValid: true,
+    }
+}
+
+export const validateRequiredPositiveNumber = (value: string | Number): ValidationResponseType => {
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Shared/validations.tsx]: Done

File: src/Pages/GlobalConfigurations/BuildInfra/services.tsx (hunk 7 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/services.tsx:65]: Done

File: src/Pages/GlobalConfigurations/BuildInfra/utils.tsx (hunk 10 above)
Comment (AbhishekA1509) [reply src/Pages/GlobalConfigurations/BuildInfra/utils.tsx:173]: Had given it a thought but issue was to maintain purity of those functions

File: src/Common/SearchBar/SearchBar.component.tsx [hunk 16]
```diff
@@ -55,21 +55,38 @@ const SearchBar = ({
...
+    }
+
     const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
         const { value } = e.target
         setShowClearButton(!!value)
+
+        if (!value) {
+            clearSearch()
+            return
+        }
```
[the remark this answers is not in the document]
Comment (AbhishekA1509) [reply src/Common/SearchBar/SearchBar.component.tsx:81]: Synced with him, its a requirement.

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**